### PR TITLE
Partial update of track IDs

### PR DESCRIPTION
**Pull request by yfukai**
This is a fresh implementation of #151. In this PR, `assign_track_ids` takes the `node_ids` argument as 
```python
rx_graph.assign_track_ids(node_ids=node_ids)
```

That only updates the track IDs for a subset of nodes connected to nodes specified by `node_ids` while maintaining the overall consistency of the ID assignment.

### CODE REVIEW

File: src/tracksdata/graph/_test/test_graph_backends.py [hunk 1]
```diff
@@ -1414,6 +1411,110 @@ def test_assign_track_ids(graph_backend: BaseGraph):
...
+            if track_id not in assigned:
+                assigned[track_id] = []
+            assigned[track_id].append(node_id)
+        assigned = {frozenset(group) for group in assigned.values()}
+        expected = {frozenset(group) for group in expected}
+        assert assigned == expected
+        assert isinstance(tracks_graph, rx.PyDiGraph)
+
+    # Check that re-assigning track IDs without reset works as expected
+    graph_backend.update_node_attrs(attrs={DEFAULT_ATTR_KEYS.TRACK_ID: -1})
```
Comment (JoOkuma) [inline src/tracksdata/graph/_test/test_graph_backends.py:1496]:
```suggestion
```
Should we remove this line because of your comment says `without reset`?

File: src/tracksdata/graph/_base_graph.py [hunk 2]
```diff
@@ -1092,6 +1092,82 @@ def bbox_spatial_filter(
...
+        Parameters
+        ----------
+        seeds : list[int]
+            Seed node IDs where to start the closure.
+
+        Returns
+        -------
+        list[int]
+            Sorted unique node IDs forming the closure.
+        """
```
Comment (JoOkuma) [inline src/tracksdata/graph/_base_graph.py:1143]:
```suggestion
        """
        # NOTE: if this function becomes a bottleneck in the future it might be worth having a specialized version per backend
```

I'm adding this comment for the future, if we ever have any issues.
Not that there's any problem with your implementation, is just that we have additional overhead from the way we materialize the successors and precessors.

File: src/tracksdata/graph/_base_graph.py [hunk 3]
```diff
@@ -1092,6 +1092,82 @@ def bbox_spatial_filter(
...
+        Returns
+        -------
+        rx.PyDiGraph
+            A compressed graph (parent -> child) with track ids lineage relationships.
+            If node_ids is provided, it will only include linages including those nodes.
+        """
+        raise NotImplementedError(f"{self.__class__.__name__} backend does not support track id assignment.")
+
+    # Internal utility to compute the exact set of nodes to assign when a subset is requested
+    def _compute_track_node_ids(self, seeds: list[int] | None) -> list[int]:
```
Comment (JoOkuma) [inline src/tracksdata/graph/_base_graph.py:1127]: I really like this function, I can see some use cases besides updating `track_ids`
Do you think we could rename and remove the `_` prefix to make it "public"?

I'm not really good at naming functions. I thought of `non_branching_connected_nodes` or `non_branching_connected_components`, but I'm open to your suggestions.

File: src/tracksdata/graph/_rustworkx_graph.py [hunk 4]
```diff
@@ -1086,48 +1086,74 @@ def assign_track_ids(
...
+                self.update_node_attrs(attrs={output_key: -1})
+                previous_id_df = None
+                if track_id_offset is None:
+                    track_id_offset = 1
+            else:
+                previous_id_df = self.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, output_key])
+                if track_id_offset is None:
+                    existing_ids = previous_id_df[output_key].unique().to_list()
+                    existing_ids = [i for i in existing_ids if i != -1]
+                    track_id_offset: int = max(existing_ids, default=0) + 1
```
Comment (JoOkuma) [inline src/tracksdata/graph/_rustworkx_graph.py:1119]: Is the current version equivalent to this?
```suggestion
                    track_id_offset: int = max(previous_id_df[output_key].max(), 0) + 1
```
this way we can avoid `unique().to_list()`

File: src/tracksdata/graph/_rustworkx_graph.py [hunk 5]
```diff
@@ -1086,48 +1086,74 @@ def assign_track_ids(
...
+                merged = new_id_df.join(
+                    previous_id_df,
+                    left_on=DEFAULT_ATTR_KEYS.NODE_ID,
+                    right_on=DEFAULT_ATTR_KEYS.NODE_ID,
+                    how="left",
+                ).filter(pl.col(output_key) != -1)
+                if merged.height > 0:
+                    track_id_map = merged.unique(output_key + "_new", keep="first")
+                    track_id_map = dict(zip(track_id_map[output_key + "_new"], track_id_map[output_key], strict=True))
+                else:
```
Comment (JoOkuma) [inline src/tracksdata/graph/_rustworkx_graph.py:1146]: This is supercool.

I'm thinking of some edge cases. If there's an update where a `tracklet` 1 is split into `tracklet` 2 and 3, will this be OK? Or would this result in `track_id_map = {2: 1, and 3:1}`?

If this is the case, you may need to filter by `merged.group_by(output_key).len()` to ensure that you're not merging when remapping.

Maybe this could also happen the other way around, where a `tracklet` 2 and 3 got fused into `tracklet` 4 and it might remap back to 2 and 3.

File: src/tracksdata/graph/_test/test_graph_backends.py [hunk 6]
```diff
@@ -1414,6 +1411,110 @@ def test_assign_track_ids(graph_backend: BaseGraph):
...
+    ids_df_reassign = ids_df_reassign.sort(DEFAULT_ATTR_KEYS.NODE_ID)
+    assert ids_df.equals(ids_df_reassign)
+
+    # Changing the topology
+    A4 = graph_backend.add_node({DEFAULT_ATTR_KEYS.T: 1, DEFAULT_ATTR_KEYS.TRACK_ID: -1})
+    graph_backend.remove_edge(A2, A3)
+    graph_backend.add_edge(A0, A4, {})
+    tracks_graph_reassign = graph_backend.assign_track_ids(node_ids=[A1, A4, B4], reset=False)
+    ids_df_reassign = graph_backend.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, DEFAULT_ATTR_KEYS.TRACK_ID])
+    assert tracks_graph_reassign.num_nodes() == 3
```
Comment (JoOkuma) [inline src/tracksdata/graph/_test/test_graph_backends.py:1516]: This test covers one of the edge cases I mentioned below.

**Review by JoOkuma — CHANGES_REQUESTED** [review]
Hi @yfukai , thanks for the PR.
I really liked the part of trying to keep the same track ids.

I left some minor comments and a few questions.

File: src/tracksdata/graph/_test/test_graph_backends.py (hunk 6 above)
Comment (yfukai) [reply src/tracksdata/graph/_test/test_graph_backends.py:1529]: Right, I added a test and it was failing with the original impl.

File: src/tracksdata/graph/_rustworkx_graph.py (hunk 5 above)
Comment (yfukai) [reply src/tracksdata/graph/_rustworkx_graph.py:1144]: Hi, thank you! I'm happy to find you liked this implementation. Also thank you for finding this issue. I missed one extra `unique` function call. I believe it's fine now, but happy to have your feedback!
The later case wouldn't happen I believe (since the "4->3" row will be dropped with "4->2" existing), but I explicitly tested that [here](https://github.com/yfukai/tracksdata/blob/90fcb7d99c36bb44dbdd9788355998f79e1e0df4/src/tracksdata/graph/_test/test_graph_backends.py).

File: src/tracksdata/graph/_test/test_graph_backends.py (hunk 6 above)
Comment (yfukai) [reply src/tracksdata/graph/_test/test_graph_backends.py:1529]: And now passing with the fixed one!

**Review by yfukai — COMMENTED** [review]
Hi, thank you @JoOkuma for your review! I'm happy to hear you liked the implementation, which I also like 👍 
I believe I addressed the issues you found, and I'm happy to have your feedback again!

**Comment by yfukai** [comment]
I updated the function name! I don't have the merge right but please feel free to go ahead 🙏

**Comment by JoOkuma** [comment]
I forgot about that again. Thanks!
